### PR TITLE
.net changes to re-enable xlang support, add subscription apis

### LINKED ISSUE #4154: add Subscription management apis to service
### What feature would you like to be added?

need to add support for 
message TypeSubscription {
    string topic_type = 1;
    string agent_type = 2;
}

message Subscription {
    oneof subscription {
        TypeSubscription typeSubscription = 1;
    }
}

message AddSubscriptionRequest {
    string request_id = 1;
    Subscription subscription = 2;
}

message AddSubscriptionResponse {
    string request_id = 1;
    bool success = 2;
    optional string error = 3;
}

### Why is this needed?

xlang combat

### PR DESCRIPTION
**Pull request by rysweet**
adding subscription request/response api
adds response on register agent 
closes #4154 
cleanup some package versions
improves hello samples
fixes a few asynhronous await bugs found along the way

### CODE REVIEW

File: dotnet/samples/Hello/Backend/Backend.csproj [hunk 1]
```diff
@@ -1,10 +1,10 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk.Web">
   <ItemGroup>
     <ProjectReference Include="..\..\..\src\Microsoft.AutoGen\Agents\Microsoft.AutoGen.Agents.csproj" />
   </ItemGroup>
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net8.0</TargetFramework>
+        <TargetFramework>net8.0</TargetFramework>
```
Comment (LittleLittleCloud) [inline dotnet/samples/Hello/Backend/Backend.csproj:7]: Formatting whitespace?

File: dotnet/samples/Hello/Hello.AppHost/Hello.AppHost.csproj [hunk 2]
```diff
@@ -1,7 +1,10 @@
 <Project Sdk="Microsoft.NET.Sdk">
+
+  <Sdk Name="Aspire.AppHost.Sdk" Version="9.0.0" />
```
Comment (LittleLittleCloud) [inline dotnet/samples/Hello/Hello.AppHost/Hello.AppHost.csproj:3]: It's something new to me, what's it for

File: dotnet/src/AutoGen.Core/Function/FunctionAttribute.cs [hunk 3]
```diff
@@ -98,7 +98,7 @@ public static implicit operator AIFunctionMetadata(FunctionContract contract)
                 [NamespaceKey] = contract.Namespace,
                 [ClassNameKey] = contract.ClassName,
             },
-            Parameters = [.. contract.Parameters?.Select(p => (AIFunctionParameterMetadata)p)],
+            Parameters = [.. contract.Parameters?.Select(p => (AIFunctionParameterMetadata)p)!],
```
Comment (LittleLittleCloud) [inline dotnet/src/AutoGen.Core/Function/FunctionAttribute.cs:101]: Is this change needed?

File: dotnet/src/Microsoft.AutoGen/Agents/Services/HostBuilderExtensions.cs [hunk 4]
```diff
@@ -28,12 +28,12 @@ public static IHostApplicationBuilder AddAgent<
     public static IHostApplicationBuilder AddAgent(this IHostApplicationBuilder builder, string typeName, Type agentType)
     {
         builder.Services.AddKeyedSingleton("AgentTypes", (sp, key) => Tuple.Create(typeName, agentType));
-
         return builder;
     }
 
-    public static IHostApplicationBuilder AddAgentWorker(this IHostApplicationBuilder builder, string agentServiceAddress = _defaultAgentServiceAddress, bool local = false)
+    public static IHostApplicationBuilder AddAgentWorker(this IHostApplicationBuilder builder, string? agentServiceAddress = null, bool local = false)
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Agents/Services/HostBuilderExtensions.cs:34]: maybe still make agentServiceAddress  required and assgin default value to it?

File: dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/SubscriptionsGrain.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SubscriptionsGrain.cs
+
+namespace Microsoft.AutoGen.Agents;
+
+public sealed class SubscriptionsGrain([PersistentState("state", "PubSubStore")] IPersistentState<SubscriptionsState> state) : Grain, ISubscriptionsGrain
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/SubscriptionsGrain.cs:6]: Does this class need to be public, can we make it internal?

File: dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/iSubscriptionsGrain.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// iSubscriptionsGrain.cs
+
+namespace Microsoft.AutoGen.Agents;
+public interface ISubscriptionsGrain : IGrainWithIntegerKey
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/iSubscriptionsGrain.cs:5]: Can we make it internal?

Comment (LittleLittleCloud) [reply dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/iSubscriptionsGrain.cs:5]: Also, can you capitalize the first character of this file name

`iSubscriptionsGrain.cs -> ISubscriptionsGrain.cs`

File: dotnet/src/Microsoft.AutoGen/Agents/Services/Grpc/GrpcGateway.cs [hunk 7]
```diff
@@ -33,6 +36,7 @@ public GrpcGateway(IClusterClient clusterClient, ILogger<GrpcGateway> logger)
         _clusterClient = clusterClient;
         _reference = clusterClient.CreateObjectReference<IGateway>(this);
         _gatewayRegistry = clusterClient.GetGrain<IRegistryGrain>(0);
+        _subscriptions = clusterClient.GetGrain<ISubscriptionsGrain>(0);
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Agents/Services/Grpc/GrpcGateway.cs:39]: Are we expecting the singeliton for this class

File: dotnet/src/Microsoft.AutoGen/Abstractions/IAgentRuntime.cs [hunk 8]
```diff
@@ -13,6 +13,7 @@ public interface IAgentRuntime
     ValueTask<AgentState> ReadAsync(AgentId agentId, CancellationToken cancellationToken = default);
     ValueTask SendResponseAsync(RpcRequest request, RpcResponse response, CancellationToken cancellationToken = default);
     ValueTask SendRequestAsync(IAgentBase agent, RpcRequest request, CancellationToken cancellationToken = default);
+    ValueTask SendMessageAsync(Message message, CancellationToken cancellationToken = default);
```
Comment (LittleLittleCloud) [inline dotnet/src/Microsoft.AutoGen/Abstractions/IAgentRuntime.cs:16]: Not related to this PR, but I wonder how close we are to combine `IAgentRuntime` and `IAgentWorker`, they looks so similar

image

File: dotnet/samples/Hello/Backend/Backend.csproj (hunk 1 above)
Comment (LittleLittleCloud) [reply dotnet/samples/Hello/Backend/Backend.csproj:7]: Same below

File: dotnet/samples/Hello/Hello.AppHost/Hello.AppHost.csproj (hunk 2 above)
Comment (rysweet) [reply dotnet/samples/Hello/Hello.AppHost/Hello.AppHost.csproj:3]: https://learn.microsoft.com/en-us/dotnet/aspire/get-started/aspire-overview

File: dotnet/samples/Hello/Backend/appsettings.Development.json [hunk 9]
```diff
@@ -0,0 +1,15 @@
+{
```
[the remark this answers is not in the document]
Comment (rysweet) [reply dotnet/samples/Hello/Backend/appsettings.Development.json:1]: done

File: dotnet/src/AutoGen.Core/Function/FunctionAttribute.cs (hunk 3 above)
Comment (rysweet) [reply dotnet/src/AutoGen.Core/Function/FunctionAttribute.cs:101]: I don't know where that one came from. I'll back it out.

Comment (rysweet) [reply dotnet/src/AutoGen.Core/Function/FunctionAttribute.cs:101]: hmm, mine won't build without it now. "Dereference of possibly null...."

File: dotnet/src/Microsoft.AutoGen/Agents/Services/Grpc/GrpcGateway.cs (hunk 7 above)
Comment (rysweet) [reply dotnet/src/Microsoft.AutoGen/Agents/Services/Grpc/GrpcGateway.cs:39]: managed by Orleans

File: dotnet/src/Microsoft.AutoGen/Agents/Services/HostBuilderExtensions.cs (hunk 4 above)
Comment (rysweet) [reply dotnet/src/Microsoft.AutoGen/Agents/Services/HostBuilderExtensions.cs:34]: this lets use use the environment var if its not set

File: dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/iSubscriptionsGrain.cs (hunk 6 above)
Comment (rysweet) [reply dotnet/src/Microsoft.AutoGen/Agents/Services/Orleans/iSubscriptionsGrain.cs]: this one (all the grains) are not necessarily internal?
renamed

**Review by LittleLittleCloud — APPROVED** [review]
